src/devices/src/virtio/net/device.rs

```diff
@@ -257,84 +277,94 @@ impl Net {
-                            if let GuestMemoryError::PartialBuffer { completed, .. } = e {
-                                write_count += completed;
-                                METRICS.net.rx_partial_writes.inc();
-                            } else {
-                                METRICS.net.rx_fails.inc();
-                            }
-                            break;
-                        }
-                    };
+            if !descriptor.is_write_only() {
+                result = Err(FrontendError::ReadOnlyDescriptor);
+                break;
+            }
 
-                    if write_count >= self.rx_bytes_read {
-                        break;
-                    }
-                    next_desc = desc.next_descriptor();
+            let len = std::cmp::min(frame_slice.len(), descriptor.len as usize);
+            match mem.write_slice(&frame_slice[..len], descriptor.addr) {
+                Ok(()) => {
+                    METRICS.net.rx_count.inc();
+                    frame_slice = &frame_slice[len..];
                 }
-                None => {
-                    warn!("Receiving buffer is too small to hold frame of current size");
-                    METRICS.net.rx_fails.inc();
+                Err(e) => {
+                    error!("Failed to write slice: {:?}", e);
+                    match e {
+                        GuestMemoryError::PartialBuffer { .. } => &METRICS.net.rx_partial_writes,
+                        _ => &METRICS.net.rx_fails,
+                    }
+                    .inc();
+                    result = Err(FrontendError::GuestMemory(e));
                     break;
                 }
-            }
-        }
+            };
 
-        rx_queue
-            .add_used(mem, head_index, write_count as u32)
-            .unwrap_or_else(|e| {
-                error!("Failed to add available descriptor {}: {}", head_index, e);
-            });
+            maybe_next_descriptor = descriptor.next_descriptor();
+        }
+        if result.is_ok() && !frame_slice.is_empty() {
+            warn!("Receiving buffer is too small to hold frame of current size");
+            METRICS.net.rx_fails.inc();
+            result = Err(FrontendError::DescriptorChainTooSmall);
+        }
 
-        // Mark that we have at least one pending packet and we need to interrupt the guest.
+        // Mark the descriptor chain as used. If an error occurred, skip the descriptor chain.
+        let used_len = result.as_ref().map_or(0, |_| frame.len() as u32);
+        queue.add_used(mem, head_index, used_len).map_err(|e| {
+            error!("Failed to add available descriptor {}: {}", head_index, e);
+            FrontendError::AddUsed(e)
+        })?;
         self.rx_deferred_irqs = true;
 
-        if write_count >= self.rx_bytes_read {
-            METRICS.net.rx_bytes_count.add(write_count);
+        if let Ok(()) = result {
+            METRICS.net.rx_bytes_count.add(frame.len());
```

Comment:
Are we always able to send the entire frame? I think we will miss deferring frames this way.

Reply:
Yes, we should always be able to send the entire frame. Any Rx descriptor chain should be able to fit an entire frame. If not, it's a guest bug. We shouldn't miss deferring frames. I'm not completely sure what you mean. Could you expand on that please ?

Reply:
My understanding is that the previous version of the code sends the rx received data even if the avail descriptors couldn't hold it entirely. If this is the case, the deferred frame is set to true, which means that the priority now is to send the rest of the rx frame instead of processing incoming data from the tap. You can observe this pattern throughout the code - e.g `process_tap_rx_event` or `handle_deffered_frame`.

> RX descriptor chain should be able to fit an entire frame, and if not, it's a guest bug. 

Do you have references about this certainty? I was not aware about it, and haven't found yet any reference myself.

Reply:
Now the behavior is the same. If the frame can't be sent, the descriptor chain is skipped, and the `deferred_frame` flag is set to true, cecause `write_frame_to_guest` returns `false`. This logic didn't change.

Quoting from the virtio standard:

```
5.1.6.3 Setting Up Receive Buffers

It is generally a good idea to keep the receive virtqueue as fully populated as possible: if it runs out, network performance will suffer.

If the VIRTIO_NET_F_GUEST_TSO4, VIRTIO_NET_F_GUEST_TSO6 or VIRTIO_NET_F_GUEST_UFO features are used, the maximum incoming packet will be to 65550 bytes long (the maximum size of a TCP or UDP packet, plus the 14 byte ethernet header), otherwise 1514 bytes. The 12-byte struct virtio_net_hdr is prepended to this, making for 65562 or 1526 bytes.
5.1.6.3.1 Driver Requirements: Setting Up Receive Buffers

    If VIRTIO_NET_F_MRG_RXBUF is not negotiated:
        If VIRTIO_NET_F_GUEST_TSO4, VIRTIO_NET_F_GUEST_TSO6 or VIRTIO_NET_F_GUEST_UFO are negotiated, the driver SHOULD populate the receive queue(s) with buffers of at least 65562 bytes.
        Otherwise, the driver SHOULD populate the receive queue(s) with buffers of at least 1526 bytes.
    If VIRTIO_NET_F_MRG_RXBUF is negotiated, each buffer MUST be at least the size of the struct virtio_net_hdr.

Note: Obviously each buffer can be split across multiple descriptor elements.

If VIRTIO_NET_F_MQ is negotiated, each of receiveq1…receiveqN that will be used SHOULD be populated with receive buffers. 
```